Documentation changes for definite assignment (Issue #3646) (#3650)
dafny-lang/dafny: docs/DafnyRef/Topics.md

## 12.6. Variable Initialization and Definite Assignment {#sec-definite-assignment}

The Dafny language semantics ensures that any use (read) of a variable (or constant,
parameter, object field, or array element) gives a value of the variable's type.
It is easy to see that this property holds for any variable that is declared with
an initializing assignment. However, for many useful programs, it would be too strict
to require an initializing assignment at the time a variable is declared.
Instead, Dafny ensures the property through _auto-initialization_ and rules for _definite assignment_.

> As explained in section 5.3.1, each type in Dafny is one of the following:

Author: Added the link

know that a value exists. Therefore, in a ghost context, a variable of nonempty type can be used immediately
after declaration.

Before a variable of a possibly empty type can be used, the program must initialize it.
The variable need not be given a value when it is declared,
but it must have a value by the time it is first used. Dafny uses the precision of the verifier to
reason about the control flow between assignments and uses of variables, and it reports an error
if it cannot assure itself that the variable has been given a value.

The elements of an array must be assured to have values already in the statement that allocates the array.
This is achieved in any of the following four ways:
- If the array is allocated to be empty (that is, one of its dimensions is required to be 0), then
  the array allocation trivially satisfies the requirement.
- If the element type of the array is an auto-init type, then nothing further is required by the program.
- If the array allocation occurs in a ghost context and the element type is a nonempty type, then nothing
  further is required by the program.

> - Otherwise, the array allocation must provide an initialization display or an initialization function.

Author: Added a link

by using an assignment statement `x := *;`, provided the type of `x` is an auto-init type (or, if `x` is
ghost, a nonempty type). (If `x` is of a possibly nonempty type, then `x := *;` is still allowed, but it
sets `x` to a value of its type only if the type actually contains a value. Therefore, when `x` is of
a possibly empty type, `x := *;` does not count as a definite assignment to `x`.)

Note that auto-initialization is nondeterministic. Dafny only guarantees that each value it assigns to
a variable of an auto-init type is _some_ value of the type. Indeed, a variable may be auto-initialized
to different values in different runs of the program or even at different times during the same run of
the program. In other words, Dafny does not guarantee the "zero-equivalent value" initialization that
some languages do. Along these lines, also note that the `witness` value provided in some subset-type
declarations is not necessarily the value chosen by auto-initialization, though it does esstablish that
the type is an auto-init type.

In some programs (for example, in some test programs), it is desirable to avoid nondeterminism.

> For that purpose, Dafny provides a `--enforce-determinism` option. It forbids use of any program

Reviewer: ```suggestion
For that purpose, Dafny provides an `--enforce-determinism` option. It forbids use of any program
```

statement that may have nondeterministic behavior and it disables auto-initialization.

> Since this mode enforces definite assignments everywhere, it forbids the use of `iterator` declarations

Reviewer: ```suggestion
This mode enforces definite assignments everywhere, going beyond what the strict mode does by enforcing
definite assignment also for fields and array elements. It also forbids the use of `iterator` declarations
```

after declaration.

Before a variable of a possibly empty type can be used, the program must initialize it.
The variable need not be given a value when it is declared,
but it must have a value by the time it is first used. Dafny uses the precision of the verifier to
reason about the control flow between assignments and uses of variables, and it reports an error
if it cannot assure itself that the variable has been given a value.

The elements of an array must be assured to have values already in the statement that allocates the array.
This is achieved in any of the following four ways:
- If the array is allocated to be empty (that is, one of its dimensions is requested to be 0), then
  the array allocation trivially satisfies the requirement.
- If the element type of the array is an auto-init type, then nothing further is required by the program.
- If the array allocation occurs in a ghost context and the element type is a nonempty type, then nothing
  further is required by the program.
- Otherwise, the array allocation must provide an initialization display or an initialization function.

> See [section 0.0.0](#sec-array-type) for information about array initialization.

Reviewer: ```suggestion
See [section 5.10](#sec-array-type) for information about array initialization.
```

Author: Oops- rerunning the auto-numbering script slipped my mind...
